File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/javascript/find_shortest_string.js
function findShortestString(arr) {
  // type your code here
  //console.log(arr)
  let srted = arr.map(a=> a.length)
  let min = Math.min(...srted)
  let idx = srted.indexOf(min)
  //console.log(idx,srted)
  return arr[idx]
}
//   let srt = arr.sort((a,b) => {
//     if (a.length < b.length)
//     return -1
//     if(a.length > b.length){
//       return 1
//      } else { 
//       return 0 
//      }
//     }
//   )
//   console.log(srt)
//   return srt[0]
// }

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  // BENCHMARK HERE
  let longArr = []
  for (let i=0; i<100;++i){
    longArr.push(Math.random().toString())
  }
  let start_time = Date.now()

  for (i=0; i<1000;++i){
    findShortestString(['a','ba'])
    findShortestString(longArr)
  }
  let avg_time = (Date.now() - start_time)/2000
  console.log('runtime',avg_time)
}

module.exports = findShortestString;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/04-day-5--find-shortest-string/ruby/find_shortest_string.rb
def find_shortest_string(arr)
  # type your code in here
  to_ints = arr.collect do |a| a.length
  end
  mind = to_ints.min
  idx = to_ints.index(mind)
  #puts mind, to_ints
  arr[idx]
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  # Don't forget to add your own!

  # BENCHMARK HERE
  ol_test = []
  100.times do 
    ol_test << rand.to_s
  end
  srt_time = Time.now

  1000.times do 
    find_shortest_string(['a','ab'])
    find_shortest_string(ol_test)
  end
  avg_time = (Time.now - srt_time)/2000
  print'avg time: ',avg_time
  
end

# Please add your pseudocode to this file
# And a written explanation of your solution
